### DISCUSSION TITLE
How to update only subject without affecting other options

### DISCUSSION THREAD

**Discussion by aniketgm** [VladimirMarkelov/ttdl#94] [General]
Let's say I add a todo as below (first image) (this is just a funny example, sorry I couldn't think of anything else while writing this):
Clearly, I made a mistake in the subject. If I try to edit it (second image), It deletes all the other options. I have to then write the whole thing (project, context, etc..) again with the updated subject. Do we have an option to edit only the subject that I'm not aware. I checked the docs and I couldn't find it.

If not, can we have an option like `--set-subj` ?

![image](https://github.com/VladimirMarkelov/ttdl/assets/14215698/a97478f7-de5d-4360-9146-6b61cb7455f1)

![image](https://github.com/VladimirMarkelov/ttdl/assets/14215698/a33c2d1e-27ee-45a7-bb25-b01beff42c11)

**Comment by VladimirMarkelov** [comment]
As far as I remember, the todo.txt format calls the entire string, including all the tags, a subject. It makes the option `--set-subj` confusing as it changes only a part of a subject. An option name like `--keep-tags` seems better to me.

It is not too hard to implement in case of all the tags are trailing or heading. But what to do if tags are scattered across the subject? Examples:

```
1 @workout at cafe due:2024-01-01
2 Buy a present for +someone birthday
```

What should happen in case of `ttdl edit 1 "at gym" --keep-tags` and `ttdl edit 2 "withdrawn 20$ for birthday" --keep-tags`? One possible solution is just move all tags to the end of the subject. Though, it can be harmful for some todos. E.g, the first todo turns into `at gym @workout due:2024-01-01`. In this simple case it does not look to bad, but there can be worse cases.

**Reply by aniketgm** [reply]
Yes, `--keep-tags` seems like a good option for now.
I forgot about the discussion we had last time https://github.com/VladimirMarkelov/ttdl/discussions/84. Users can do their own customization and make it work.

A different solution came into my mind. But that might be a big change. Your idea of writing todo is that user should feel writing todo's as natural as possible `ttdl Visit @walmart on due:2024-04-04 to buy groceries`, instead of `ttdl Buy groceries @walmart due:2024-04-04`

Unless we use some AI, it will be difficult to edit and maintain such todo string due to the complexity. One solution is to present the user the whole string. User edits the string, saves it and the saved string is updated in the todo.txt.

By **"present the user the whole string"**, I mean opening the string in some editor of choice. I use vim, however not all would be using it. Some might like notepad++, etc. And this might be a big change.

A simple solution would be to present the `--keep-tags` with a _disclaimer_ to the users that, the order of words will change when used. And as suggested above, the `--keep-tags` would move all tags to the end of the subject.

**Reply by VladimirMarkelov** [reply]
Taking into account that 1) replacing only the subject seems a rare case as nobody asked for it before and 2) using tags inside a subject is also a rare thing, I guess that the simple solution that you have mentioned above - just put all tags at the end - would work in vast majority of cases. Maybe, to make it easier to use, we should also introduce a short option, e.g, `-k`, so it would be faster to type (typing `--keep-tags` is not much easier that typing `@walmart` 😄  ).

Also, I guess usually people want to edit subject almost right after adding a new task because they notice a typo or alike. In this case, instead of editing, it maybe faster and easier just get the last command from shell history, modify it and add a new task or edit the exiting one.

**Comment by VladimirMarkelov** [comment]
The task is in the progress. The coding part is done, documentation is left.
